docs/developers/user-data.md

```diff
@@ -1,11 +1,98 @@
 [![PyPI version](https://badge.fury.io/py/dsp-tools.svg)](https://badge.fury.io/py/dsp-tools)
 
-# User data in the folder `.dsp-tools`
+# User data in the user's home directory
 
-DSP-TOOLS saves user data in the user's home directory, in the folder `.dsp-tools`. Here is an overview of its 
-structure:
+DSP-TOOLS saves user data in the user's home directory, 
+in the folder `.dsp-tools`. 
+Here is an overview of its structure:
 
 | folder     | command using it | description                                  |
 |:-----------|:-----------------|:---------------------------------------------|
 | xmluploads | `xmlupload`      | saves id2iri mappings and error reports      |
 | docker     | `start-stack`    | files necessary to startup Docker containers |
+
+Remark: Docker is normally not able to access files 
+stored in the `site-packages` of a Python installation.
+Therefore, it's necessary to copy the "docker" folder
+to the user's home directory.
+
+
+
+## How to ship data files to the user
+
+Accessing non-Python files (a.k.a. resources, a.k.a data files) 
+in the code needs special attention.
+
+Firstly, the build tool must be told to include this folder/files in the distribution.
+In our case, this happens in `[tool.poetry.include]` in the `pyproject.toml` file.
+
+Secondly, when accessing the files on the customer's machine, 
+the files inside `site-packages` should be read-only 
+to avoid a series of common problems 
+(e.g. when multiple users share a common Python installation, 
+when the package is loaded from a zip file, 
+or when multiple instances of a Python application run in parallel).
+
+Thirdly, the files can neither be accessed 
+with a relative path from the referencing file,
+nor with a path relative to the root of the project.
+
+For example, if you have a structure like this:
+
+```
+dsp-tools
+├── pyproject.toml
+└── src
+    └── dsp_tools
+        ├── schemas
+        │   └── data.xsd
+        ├── __init__.py
+        └── dsp_tools.py
+```
+
+it is not possible to do one of the following in dsp_tools/dsp_tools.py:
+
+```python
+with open('schemas/data.xsd') as data_file:
+     ...
+with open('src/dsp_tools/schemas/data.xsd') as data_file:
+     ...
+```
+
+The reason why these two approaches fail is 
+that the working directory on the user's machine 
+is determined by the directory where 
+DSP-TOOLS is called from - 
+not the directory where the distribution files are situated in.
+
+To circumvent this problem,
+it was once common to manipulate a package’s `__file__` attribute 
+in order to find the location of data files:
+
+```python
+import os
+data_path = os.path.join(os.path.dirname(__file__), 'schemas', 'data.xsd')
+with open(data_path) as data_file:
+     ...
+```
+
+However, this manipulation isn’t compatible with PEP 302-based import hooks, 
+including importing from zip files and Python Eggs.
+
```

Comment:
Once more, I don't think that explanations how things were in earlier time should be included in documentation. I would prefer software documentation to be consise.

Reply:
I thought that the "developers" part of the docs are here for exactly this... ? I created this part of the docs to write all the stuff there that I would not write into the normal documentation. I thought that the "developers" part can be more reflective, with historical things, with reasons why we do some things... Or did I get that wrong?

Reply:
Of course, it very much depends on the goal of the documentation and the intended audience. But to me, documentation should only be about _how_ to do things. The _why_ should go into decision records. If I would need to read through paragraphs about how things were done previously - it would cost me time and effort without getting to the relevant information that I actually need to do something.

Also, the _why_ is important for big decisions (e.g. architecture) and not so much small changes. From this perspective, the way to open a file is an implementation detail. If you would describe every change in this level of detail, you would spend a lot of time writing documentation :) But maybe I don't fully understand you point of view: so, I would be curious to hear what you think. Why does a developer need to know how files were opened previously?